horovod/spark/common/util.py

```diff
@@ -0,0 +1,518 @@
+                size = shape
+            elif dtype == SparseVector:
+                # shape is the total size of vector
+                shape = data_col.size
+                # size is the number of nonzero elements in the sparse vector
+                size = data_col.indices.shape[0]
+            else:
+                shape = 1
+                size = 1
+            row_schema.append((col_name, (set([dtype]), set([shape]), set([size]))))
+        return row_schema
+
+    def merge(x, y):
+        dtypes = x[0]
+        dtypes.update(y[0])
+        shapes = x[1]
+        shapes.update(y[1])
+        sizes = x[2]
+        sizes.update(y[2])
+        return dtypes, shapes, sizes
+
+    raw_col_info_list = df.rdd.flatMap(get_meta).reduceByKey(merge).collect()
+
+    all_col_types = {}
+    col_shapes = {}
+    col_max_sizes = {}
+
+    for col_info in raw_col_info_list:
+        col_name = col_info[0]
+
+        all_col_types[col_name] = col_info[1][0]
+        col_shapes[col_name] = col_info[1][1]
+        col_max_sizes[col_name] = col_info[1][2]
+
+    # all the rows of each columns must have the same shape
+    for col in df.schema.names:
+        shape_set = col_shapes[col]
+        if len(shape_set) != 1:
+            raise ValueError(
+                'col {col} does not have uniform shapes. shape set: {shapes_set}'.format(col=col,
+                                                                                         shapes_set=shape_set))
+        col_shapes[col] = shape_set.pop()
+
+    for col in df.schema.names:
+        sizes = col_max_sizes[col]
+        if len(sizes) > 1 and not (SparseVector in all_col_types[col]):
+            raise ValueError(
+                "rows of column {col} have varying sizes. This is only allowed if datatype is "
+                "SparseVector or a mix of Sparse and DenseVector.".format(col=col))
+        col_max_sizes[col] = max(sizes)
+
+    return all_col_types, col_shapes, col_max_sizes
+
+
+def _get_metadata(df):
+    """
+    Infer the type and shape of all the columns and determines if what intermedite format they
+    need to be converted to in case they are a vector.
+
+    Example return value:
+    {
+    'col1': {
+        'dtype': <type 'float'>,
+        'intermediate_format': 'nochange',
+        'max_size': 1,
+        'shape': 1
+        },
+     'col2': {
+        'dtype': <type 'float'>,
+        'intermediate_format': 'nochange',
+        'max_size': 1,
+        'shape': 1
+        },
+     'col3': {
+        'dtype': <class 'pyspark.ml.linalg.SparseVector'>,
+        'intermediate_format': 'custom_sparse_format',
+        'max_size': 37,
+        'shape': 56
+        }
+    }
+    """
+    all_col_types, col_shapes, col_max_sizes = _get_col_info(df)
+
+    metadata = dict()
+    for field in df.schema.fields:
+        col = field.name
+        col_types = all_col_types[col].copy()
+
+        if DenseVector in col_types and SparseVector in col_types:
```

Comment:
you can drop the first `if`clause, it is included in the second, because the bodies are identical:
`if dense and sparse or dense` is the same as `if dense`:
```
if DenseVector in col_types:
elif SparseVector in col_types: # and not dense
else: # neither dense nor sparse
```

Reply:
Good catch!  Fixed.